### DISCUSSION TITLE
Is there support on standard 5G NR test models for reference wave generation in Py3gpp

### DISCUSSION THREAD

**Discussion by amulamoottil** [catkira/py3gpp#31] [General]
Hi,
Is there support on standard 5G NR test models for reference wave generation in Py3gpp ?

**Comment by catkira** [comment]
not yet, but we can implement it :)
Can You describe in more detail what You are looking for?

**Comment by amulamoottil** [comment]
This example shows how to generate standard-compliant 5G NR test models (NR-TMs) and uplink and downlink fixed reference channels (FRCs) for frequency range 1 (FR1) and frequency range 2 (FR2). For the NR-TM and FRC waveform generation, you can specify the NR-TM or FRC name, the channel bandwidth, the subcarrier spacing, and the duplexing mode.

The 3GPP 5G NR standard defines sets of link and waveform configurations for the purposes of conformance testing. Two specific types of downlink conformance waveforms are NR test models (NR-TM), for the purpose of base station (BS) RF testing, and downlink fixed reference channels (FRC), for user equipment (UE) input testing.

The NR-TMs for FR1 are defined in [TS 38.141-1](https://www.3gpp.org/ftp/Specs/archive/38_series/38.141-1) Section 4.9.2, and the NR-TMs for FR2 are defined in [TS 38.141-2](https://www.3gpp.org/ftp/Specs/archive/38_series/38.141-2) Section 4.9.2.
Example is as given below

https://www.mathworks.com/help/5g/ug/5g-nr-tm-waveform-generation.html

**Reply by amulamoottil** [reply]
Hey @catkira  also  is there support to create a complex signal according to the parameters we give like ssb type, BW, SCS etc. It would be really cool if Py3gpp supports it and would be a huge step in the goal of py3gpp

**Reply by catkira** [reply]
yes, I think it would be nice to implement this waveform generator in py3gpp. Should be pretty straight forward. I don't have time for it at the moment (still busy with open5G_phy), but You can give it a try and do a PR.

**Reply by amulamoottil** [reply]
Sure will work on it.  I am trying to figure out how to do that with the available functions of Py3gpp. Would you have a spec sheet of py3gpp or any specifications as such for the same so that I can utilize the code examples so it will be faster. For example there are many constant numbers given which I am not familiar in the py3gpp example. This would be really helpful. If the same is not there its ok I can try to make it work but it will be a lot harder
